test/extensions/filters/http/aws_lambda/aws_lambda_filter_test.cc

```diff
@@ -321,6 +339,64 @@ TEST_F(AWSLambdaFilterTest, SignOnTrailedEndStream) {
   EXPECT_TRUE(headers.has("Authorization"));
 }
 
+TEST_F(AWSLambdaFilterTest, SignsDataSetByPreviousFilters) {
+  // there are cases where even if our filter is waiting on decode headers
+  // decode data can still happen. We need to verify that bodysha is still
+  // handled appropriately.
+
+  // we will manage credentials callbacks ourselves for this test
+  setupRoute(false, false, false, false, true);
+
+  Http::TestRequestHeaderMapImpl headers{{":method", "GET"},
+                                         {":authority", "www.solo.io"},
+                                         {":path", "/getsomething"}};
+  Buffer::OwnedImpl data("data");
+  Http::TestRequestTrailerMapImpl trailers;
+
+
+  StsConnectionPool::Context::Callbacks *callbackReference;
+  EXPECT_CALL(*filter_config_, getCreds).WillRepeatedly(
+     [&](StsConnectionPool::Context::Callbacks *callbacks){
+        callbackReference = callbacks;
+     })
+  );
+
+  filter_->decodeHeaders(headers, false);
+  callbackReference->onSuccess(filter_config_->credentials_);
+  filter_->decodeData(data, false);
+  filter_->decodeTrailers(trailers);
+  auto auth = filter_->aws_authenticator();
+  auto hex_sha1 = auth.getBodyHexSha();
+
+  filter_ = std::make_unique<AWSLambdaFilter>(
+      factory_context_.cluster_manager_, factory_context_.api_,
+      filter_config_);
+  filter_->setDecoderFilterCallbacks(filter_callbacks_);
+
+
+  // Pointers are a thing so lets just restate the above values
+  Http::TestRequestHeaderMapImpl headers_2{{":method", "GET"},
+                                         {":authority", "www.solo.io"},
+                                         {":path", "/getsomething"}};
+  Buffer::OwnedImpl data_2("data");
+  Http::TestRequestTrailerMapImpl trailers_2;
+
+  Buffer::OwnedImpl diff_data("different");
+
+
+
+  // intentionally misorder our on success to simulate a long sts call.
+  filter_->decodeHeaders(headers_2, false);
+  filter_->decodeData(data_2, false);
```

Comment:
nice test! is it worth adding one where onsuccess happens after `decodeData(data_2, true);` (with no decodeTrailers) to better simulate the bug we observed? (to clarify, i think it should be in addition to this case as both are valid)

Reply:
It definitely was worth it. Added it and also in the course of it correctly added a context to tell the filter that its running async to sts.